qiskit/extensions/unitary.py

```diff
@@ -62,6 +63,10 @@ def __init__(self, data, label=None):
         if input_dim != output_dim or 2**n_qubits != input_dim:
             raise ExtensionError(
                 "Input matrix is not an N-qubit operator.")
+
+        self._qasm_name = None
+        self._qasm_definiton = None
+        self._qasm_def_written = False
```

Comment:
Maybe `self._qasm_definiton` can be merged with `self._qasm_def_written`?

Reply:
I thought about this but I decided to keep them separate. `self._qasm_definition` holds the definition of this gate in QASM, and `self._qasm_def_written` keep track of whether in this call to the method we have already written the QASM. By keeping them separate we don't have to regenerate the QASM definition of the gate with every call to `circuit.qasm()` and I think this should therefore offer a slight speedup.

Reply:
I think we do similar things in other parts with of the code with `@property`. Let me write come up with an example ...

Reply:
Also, why this caching is needed here and not in the rest of the gates?

Reply:
Other gates can be written just using their name, so no caching is needed as nothing is being generated. This gate needs to be written as a new gate definition, so we have to generate the QASM definition so we use caching so we don't have to keep regenerating it every time.